byzcoin/trie/node.go

```diff
@@ -0,0 +1,165 @@
+package trie
+
+import (
+	"crypto/sha256"
+	"encoding/binary"
+	"errors"
+	"hash"
+
+	"github.com/dedis/protobuf"
+)
+
+type nodeType int
+
+const (
+	typeInterior nodeType = iota
+	typeEmpty
+	typeLeaf
+)
+
+func (ty nodeType) toBytes() []byte {
+	switch ty {
+	case typeInterior:
+		return []byte{0}
+	case typeEmpty:
+		return []byte{1}
+	case typeLeaf:
+		return []byte{2}
+	default:
+		panic("no such type")
+	}
+}
+
+type interiorNode struct {
+	Left  []byte
+	Right []byte
+}
+
+func (n *interiorNode) hash() []byte {
+	h := sha256.New()
+	h.Write(n.Left)
+	h.Write(n.Right)
+	return h.Sum(nil)
+}
+
+func (n *interiorNode) encode() ([]byte, error) {
+	buf, err := protobuf.Encode(n)
+	if err != nil {
+		return nil, err
+	}
+	return append(typeInterior.toBytes(), buf...), nil
+}
+
+func newInteriorNode(left, right []byte) interiorNode {
+	return interiorNode{
+		Left:  left,
+		Right: right,
+	}
+}
+
+func decodeInteriorNode(buf []byte) (interiorNode, error) {
+	if len(buf) == 0 {
+		return interiorNode{}, errors.New("empty buffer")
+	}
+	if nodeType(buf[0]) != typeInterior {
+		return interiorNode{}, errors.New("wrong node type")
+	}
+	var node interiorNode
+	if err := protobuf.Decode(buf[1:], &node); err != nil {
+		return interiorNode{}, err
+	}
+	return node, nil
+}
+
+type emptyNode struct {
+	Prefix []bool
+}
+
+func newEmptyNode(prefix []bool) emptyNode {
+	return emptyNode{
+		Prefix: prefix,
+	}
+}
+
+func (n *emptyNode) hash(nonce []byte) []byte {
+	h := n.hashInterface(nonce)
+	return h.Sum(nil)
+}
+
+func (n *emptyNode) hashInterface(nonce []byte) hash.Hash {
+	h := sha256.New()
+	h.Write(typeEmpty.toBytes())
```

Comment:
Style: Isn't `[]byte{typeEmpty}` also good here? I suppose we'll never have multi-byte identifiers...